Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt

```diff
@@ -17,81 +17,134 @@
     }
 
-  override fun getCount() =
-    (controller.endDate.year - controller.startDate.year) * MONTHS_IN_YEAR +
-      controller.endDate.month.value - controller.startDate.month.value + 1
+  private var positionMetadata = computeMetadata()
 
-    override fun getItem(position: Int) = null
+  init {
+    registerDataSetObserver(object : DataSetObserver() {
+      override fun onChanged() {
+        positionMetadata = computeMetadata()
+      }
+    })
+  }
 
-    override fun getItemId(position: Int) = position.toLong()
+  override fun onDayClick(view: MonthView?, day: LocalDate) {
+    onDayTapped(day)
+  }
 
-    override fun hasStableIds() = true
+  override fun getCount() = positionMetadata.size
 
-    @Suppress("UNCHECKED_CAST")
-    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
-        val view: MonthView
+  override fun getItem(position: Int) = null
 
-        if (convertView != null) {
-            view = convertView as MonthView
-        } else {
-            view = createMonthView(context).apply {
-                layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
-                isClickable = true
-            }
-            view.onDayClickListener = this@MonthAdapter
-        }
+  override fun getItemId(position: Int) = position.toLong()
 
-        val positionWithStart = position + controller.startDate.month.value - 1
-        val month = positionWithStart % MONTHS_IN_YEAR + 1
-        val year = positionWithStart / MONTHS_IN_YEAR + controller.startDate.year
-        val selectedDay = selectedDay?.let {
-          if (isSelectedDayInMonth(it, year, month)) it.dayOfMonth else null
-        }
+  override fun hasStableIds() = true
 
-        view.reuse()
+  @Suppress("UNCHECKED_CAST")
+  override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+    val metadata = positionMetadata[position]
+    val month = metadata.month
+    val year = metadata.year
+    val viewType = metadata.viewType
 
-        view.setMonthParams(CalendarDrawingParams(year, month, selectedDay, controller.calendarColoring, controller::isDateDisabled))
-        view.invalidate()
+    return if (viewType == VIEW_TYPE_FOOTER) {
+      getFooterView(convertView, month, year)
+    } else {
+      getMothView(convertView, month, year)
+    }
+  }
+
+  private fun createMonthView(context: Context) =
+    MonthView(context).also {
+      it.controller = controller
+      it.isClickable = true
+      it.onDayClickListener = this@MonthAdapter
+  }
+
+  private fun isSelectedDayInMonth(calendarDay: LocalDate, year: Int, month: Int) =
+    calendarDay.year == year && calendarDay.month.value == month
+
+  private fun onDayTapped(day: LocalDate) {
+    controller.onDayOfMonthSelected(day)
+    selectedDay = day
+  }
+
+  private fun getFooterView(convertView: View?, month: Int, year: Int): View {
+    val shouldBind = convertView != null && convertView !is MonthView
```

Comment:
As far as I remember you don't need to do a type check when you use the viewTypes. Let's just check it and make sure it works okay without it.